core/src/main/java/google/registry/tools/server/RefreshDnsForAllDomainsAction.java

```diff
@@ -76,27 +80,39 @@ public class RefreshDnsForAllDomainsAction implements Runnable {
   public void run() {
     assertTldsExist(tlds);
     checkArgument(smearMinutes > 0, "Must specify a positive number of smear minutes");
-    tm().transact(
-            () ->
-                tm().query(
-                        "SELECT domainName FROM Domain "
-                            + "WHERE tld IN (:tlds) "
-                            + "AND deletionTime > :now",
-                        String.class)
-                    .setParameter("tlds", tlds)
-                    .setParameter("now", clock.nowUtc())
-                    .getResultStream()
-                    .forEach(
-                        domainName -> {
-                          try {
-                            // Smear the task execution time over the next N minutes.
-                            requestDomainDnsRefresh(
-                                domainName, Duration.standardMinutes(random.nextInt(smearMinutes)));
-                          } catch (Throwable t) {
-                            logger.atSevere().withCause(t).log(
-                                "Error while enqueuing DNS refresh for domain '%s'.", domainName);
-                            response.setStatus(HttpStatus.SC_INTERNAL_SERVER_ERROR);
-                          }
-                        }));
+    String lastInPreviousBatch = "";
+    while (true) {
```

Comment:
Infinite while loops are dangerous and aren't all that readable, as from a moment's glance it's not clear what conditions would cause the loop to terminate.

This will be much more readable using a do-while, which would look like, in pseudo-code:

```
do {
  get new batch;
  process batch;
} while { size of latest batch == BATCH_SIZE };
```